### DISCUSSION TITLE
ServerSide Validation

### DISCUSSION THREAD

**Discussion by aamirbhat** [data-driven-forms/react-forms#1178] [Need help] [answered]
**Scope**: <DESCRIBE SCOPE OF THIS ISSUE, i.e.: PF4 mapper, Form renderer>
Form renderer
How to raise server-side validation error on submit like username already exists, The error needs to be raised on async submit after some intervals

**Description**
We need to have server-side validation on submit, those errors are raised by the backend API , we are using MUI

**Comment by rvsia** [comment, marked as answer]
Hello,

in your submit function you have to return an object with errors.

Check this example: https://stackblitz.com/edit/2qj4zm?file=index.js

```jsx
        <FormRenderer
          schema={schema}
          FormTemplate={FormTemplate}
          componentMapper={componentMapper}
          onSubmit={async () => {
            await sleep(1500); // do your API callback

            return { 'first-name': 'already taken' };
          }}
        />
```
